p2p/testing/protocolsession.go

```diff
@@ -0,0 +1,224 @@
+type Exchange struct {
+	Label    string
+	Triggers []Trigger
+	Expects  []Expect
+}
+
+// part of the exchange, incoming message from a set of peers
+type Trigger struct {
+	Msg     interface{}     // type of message to be sent
+	Code    uint64          // code of message is given
+	Peer    discover.NodeID // the peer to send the message to
+	Timeout time.Duration   // timeout duration for the sending
+}
+
+type Expect struct {
+	Msg     interface{}     // type of message to expect
+	Code    uint64          // code of message is now given
+	Peer    discover.NodeID // the peer that expects the message
+	Timeout time.Duration   // timeout duration for receiving
+}
+
+type Disconnect struct {
+	Peer  discover.NodeID // discconnected peer
+	Error error           // disconnect reason
+}
+
+// trigger sends messages from peers
+func (self *ProtocolSession) trigger(trig Trigger) error {
+	simNode, ok := self.adapter.GetNode(trig.Peer)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v does not exist (1- %v)", trig.Peer, len(self.IDs))
+	}
+	mockNode, ok := simNode.Services()[0].(*mockNode)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v is not a mock", trig.Peer)
+	}
+
+	errc := make(chan error)
+
+	go func() {
+		log.Trace(fmt.Sprintf("trigger %v (%v)....", trig.Msg, trig.Code))
+		errc <- mockNode.Trigger(&trig)
+		log.Trace(fmt.Sprintf("triggered %v (%v)", trig.Msg, trig.Code))
+	}()
+
+	t := trig.Timeout
+	if t == time.Duration(0) {
+		t = 1000 * time.Millisecond
+	}
+	alarm := time.NewTimer(t)
+	select {
+	case err := <-errc:
+		return err
+	case <-alarm.C:
+		return fmt.Errorf("timout expecting %v to send to peer %v", trig.Msg, trig.Peer)
+	}
+}
+
+// expect checks an expectation
+func (self *ProtocolSession) expect(exp Expect) error {
+	if exp.Msg == nil {
+		return errors.New("no message to expect")
+	}
+	simNode, ok := self.adapter.GetNode(exp.Peer)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v does not exist (1- %v)", exp.Peer, len(self.IDs))
+	}
+	mockNode, ok := simNode.Services()[0].(*mockNode)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v is not a mock", exp.Peer)
+	}
+
+	errc := make(chan error)
+	go func() {
+		log.Trace(fmt.Sprintf("waiting for msg, %v", exp.Msg))
+		errc <- mockNode.Expect(&exp)
+	}()
+
+	t := exp.Timeout
+	if t == time.Duration(0) {
+		t = 2000 * time.Millisecond
+	}
+	alarm := time.NewTimer(t)
+	select {
+	case err := <-errc:
+		log.Trace(fmt.Sprintf("expected msg arrives with error %v", err))
+		return err
+	case <-alarm.C:
+		return fmt.Errorf("timout expecting %v sent to peer %v", exp.Msg, exp.Peer)
+	}
+}
+
+// TestExchanges tests a series of exchanges againsts the session
+func (self *ProtocolSession) TestExchanges(exchanges ...Exchange) error {
+	// launch all triggers of this exchanges
+
+	for i, e := range exchanges {
+		errc := make(chan error)
+		wg := &sync.WaitGroup{}
+		for _, trig := range e.Triggers {
+			err := self.trigger(trig)
+			if err != nil {
+				errc <- err
```

Comment:
There is no goroutine reading `errc` here so it will block forever, should it just be `return err`?

Reply:
i made errc buffered and break after first error

Reply:
What if there are two triggers and they both fail? Is there a reason we don't just return the error?